### PR TITLE
NEO-2309: Adding Custom Filter support to Table

### PR DESCRIPTION
**Pull request by enrique-prado**
[Link to Table story in Deploy Preview](https://deploy-preview-500--neo-react-library-storybook.netlify.app/?path=/story/components-table--custom-basic-table-filter-drawer)

**Before tagging the team for a review, I have done the following:**

- [x] run `yarn all` locally: ensures that all tests pass, formatting is done, types pass, and builds pass
- [x] reviewed my code changes
- [x] updated the Deploy Preview link at the top of this PR (or remove it if not applicable)
- [x] added any important information to this PR (description, images, GIFs, ect.)
- [x] done an accessibility check on my work (tested with Chrome's `axe Dev Tools`, Mac's VoiceOver, etc.)
- [x] tagged `@avaya-dux/dux-design` if any visual changes have occurred
- [x] tagged `@avaya-dux/dux-devs`

This PR implements support for custom table filter. Devs can now override the default behavior / UI.

`@avaya-dux/dux-design`: This is just a basic example of what can be done but it's up to the developer to implement any UI design.
- first focus point: Basic example that uses the Drawer.
- second focus point: The main point of the PR is to show that you can override what happens when you click on the filter icon on the table toolbar next to the search input field.
- "not in scope":  An example of how to filter actual content will be done in Part 2 of the PR.

`@avaya-dux/dux-devs`: Created a new callback called "handleShowColumnsFilter" to override the default behavior when a user clicks on the filter button.

- handleShowColumnsFilter only gets called if it is actually provided, otherwise the default "choose columns" filter drawer is shown.
- Created Story that shows a basic custom filter drawer showcasing how to override the default behavior.

image

image

## Summary by CodeRabbit

- **New Features**
	- Enhanced reusability of the Drawer component with an updated `BaseDrawerProps` interface.
	- Introduced `CustomBasicTableFilterDrawer` to enable interactive column selection in tables.
	- Implemented `handleChooseColumns` and `handleShowColumnsFilter` props to manage column visibility in tables.
	- Added a new `TableFilterDrawer` component for a more intuitive table filtering experience.

- **Bug Fixes**
	- Improved error handling through the addition of optional method signatures across components.

- **Documentation**
	- Expanded exports in the Table component suite for enhanced modularity and improved interface clarity.

### CODE REVIEW

File: src/components/Table/TableComponents/TableFilterDrawer.tsx [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+			title,
+			translations,
+
+			...rest
+		}: TableFilterDrawerProps,
+		ref: React.Ref<HTMLDivElement>,
+	) => {
+		// translations for button labels
+		const applyLabel = translations?.apply || "Apply";
+		const cancelLabel = translations?.cancel || "Cancel";
```
Comment (joe-s-avaya) [inline src/components/Table/TableComponents/TableFilterDrawer.tsx:36]: good work remembering all the translation stuff and implementing in an simple way!

File: src/components/Table/TableComponents/TableFilterDrawer.tsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+		const actionButtons = [
+			<Button
+				form={id}
```
Comment (joe-s-avaya) [inline src/components/Table/TableComponents/TableFilterDrawer.tsx:40]: you're passing a form id but neither of the buttons are `type=submit`... sus...

**Comment by matthewAvaya** [comment]
Although you don't have any content in the middle section, clicking on the **Apply** button should dismiss the Drawer. Should this example have that behavior, even though there's no content to apply the filter?

**Review by Justin-Avaya — APPROVED** [review]
Apply button does not close the drawer. Probably out of scope.

File: src/components/Table/TableComponents/TableFilterDrawer.tsx (hunk 2 above)
Comment (enrique-prado) [reply src/components/Table/TableComponents/TableFilterDrawer.tsx:40]: it's not just for buttons type=submit. It's basically used when a button is located outside of the form but it still belongs to that form.  Having said that, there is no form tag in the drawer implementation so I'm not sure why I put it there. Probably a copy pasta oversight. Nice catch. Removed.

**Comment by enrique-prado** [comment]
> Seems pretty straightforward. I left one question but otherwise this looks quite simple.
> 
> I recommend bringing in some tests to assert that the default filter does not open when the override method is passed.

Good suggestion. I checked and we do have one that already verifies the default behavior in Table.test.jsx line 760

**Comment by enrique-prado** [comment]
> Although you don't have any content in the middle section, clicking on the **Apply** button should dismiss the Drawer. Should this example have that behavior, even though there's no content to apply the filter?

That's part of the custom behavior that the developer needs to provide. It's totally up to the developer to write the code. I simply chose to only provide code for the Cancel button.

**Pull request merged**